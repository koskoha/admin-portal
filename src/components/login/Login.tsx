import * as React from 'react';
import { graphql } from 'react-apollo';
import { gql } from 'apollo-boost';
import { Redirect } from 'react-router-dom';

import { AUTH_TOKEN } from '../../config/constant';
import Input from '../elements/Input';
import logoIcon from '../../images/logo.png';

interface LoginState {
	email: string;
	password: string;
	serverErrors: [] | undefined;
	emailError: string | undefined;
	passwordError: string | undefined;
	isLogined: boolean;
}

interface LoginProps {
	refreshTokenFn: ({ [AUTH_TOKEN]: string }) => void;
	login: ({ variables }) => void;
}

class Login extends React.Component<LoginProps, LoginState> {
	constructor(props) {
		super(props);
		this.state = {
			email: '',
			password: '',
			serverErrors: undefined,
			emailError: '',
			passwordError: '',
			isLogined: false
		};
	}

	renderErrors = errors => (
		<div className="error-wrapper">
			<ul>
				{errors.map(error => (
					<li key={error}>{error.message}</li>
				))}
			</ul>
		</div>
	);

	handlePasswordChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.setState({ password: e.currentTarget.value, passwordError: '' });
	};

	handleEmailChange = (e: React.ChangeEvent<HTMLInputElement>) => {
		this.setState({ email: e.currentTarget.value, emailError: '' });
	};

	validatePassword = (value: string): string => {
		const error = value ? '' : 'Password can not be empty!';
		this.setState({ passwordError: error });
		return error;
	};

	validateEmail = (value: string): string => {
		const error = /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/.test(
			value
		)
			? ''
			: 'You must enter a valid email address';
		this.setState({ emailError: error });
		return error;
	};

	validateSubmit = (): void => {
		const { password, email } = this.state;
		this.setState({ serverErrors: undefined });
		const passwordError = this.validatePassword(password);
		const emailError = this.validateEmail(email);
		if (passwordError === '' && emailError === '') {
			this.loginUser();
		}
	};

	render() {
		const {
			email,
			password,
			serverErrors,
			emailError,
			passwordError,
			isLogined
		} = this.state;
		return (
			<div className="login-page">
				{isLogined && <Redirect to="/dashboard/accounts" />}
				<div className="login-wrapper">
					{/* HEADER FORM */}
					<div className="login-header">
						<div>
							<img src={logoIcon} />
						</div>
						<div className="company-title">
							<div className="title">SpecialtyCare Connect</div>
							<div className="sub-title">POWERED BY ARMADAHEALTH</div>
						</div>
					</div>

					{/* BODY FORM */}
					<div className="login-form-wrapper">
						<span className="form-title">Sign in to the Admin Portal</span>
						{serverErrors && this.renderErrors(serverErrors)}
						<div className="login-form">
							<Input
								name="email"
								value={email}
								onChange={this.handleEmailChange}
								label="Email Address"
								placeholder="Email Address"
								style={{ color: '#ffff' }}
							/>
							<span className="error">{emailError}</span>
							<Input
								name="password"
								value={password}
								onChange={this.handlePasswordChange}
								label="Password"
								placeholder="Password"
								type="password"
								style={{ color: '#ffff' }}
							/>
							<span className="error">{passwordError}</span>
						</div>
						{/* FOOTER FORM */}
						<div className="login-footer">Forgot password? Click here.</div>

						<button
							onClick={this.validateSubmit}
							className="button"
							style={{ width: '100%' }}
						>
							Login
						</button>
					</div>
				</div>
			</div>
		);
	}

	loginUser = async () => {
		const { email, password } = this.state;
		const { refreshTokenFn } = this.props;
		this.props
			.login({
				variables: {
					input: {
						email,
						password
					}
				}
			})
			.then(result => {
				const token = result.data.login.accessToken;
				this.setState({ serverErrors: undefined });
				refreshTokenFn({
					[AUTH_TOKEN]: token
				});
				this.setState({ isLogined: true });
			})
			.catch(({ graphQLErrors }) => {
				this.setState({ serverErrors: graphQLErrors });
			});
	};
}

const LOGIN_USER_MUTATION = gql`
	mutation Login($input: LoginInput!) {
		login(input: $input) {
			accessToken
		}
	}
`;

export default graphql(LOGIN_USER_MUTATION, { name: 'login' })(Login);
export { LOGIN_USER_MUTATION };
