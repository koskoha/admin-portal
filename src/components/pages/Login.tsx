import * as React from 'react';
import { styled } from 'baseui';
import { Link } from 'react-router-dom';

import FormInput from '../elements/LoginInput';
import LoginButton from '../elements/Button';

const LoginForm = styled('div', {
	minWidth: '100%',
	width: '100%'
});

const Label = styled('div', {
	fontWeight: '600',
	fontStyle: 'normal',
	fontSize: '18px',
	lineHeight: '1',
	letterSpacing: '0px',
	marginTop: '15px',
	marginBottom: '30px'
});

const FormTitle = styled('span', {
	fontSize: '22px',
	fontWeight: 'bold',
	marginTop: '20px',
	marginBottom: '40px'
});

export default class Login extends React.Component {
	render() {
		return (
			<div className="login-page">
				<div className="login-form">
					{/* HEADER FORM */}
					<div className="login-form-header">
						<div>logo</div>
						<div className="company-title">
							<div className="title">SpecialtyCare Connect</div>
							<div className="sub-title">POWERED BY ARMADAHEALTH</div>
						</div>
					</div>

					{/* BODY FORM */}
					<div className="login-form-body">
						<FormTitle>Sign in to Admin Portal</FormTitle>
						<LoginForm>
							<FormInput label="Email Address" placeholder="Email Address" />
							<FormInput
								label="Password"
								type="password"
								placeholder="Password"
							/>
						</LoginForm>
						<Label>
							Forgot password?
							<Link className="link" to="/admin">
								Click here.
							</Link>
						</Label>
						<Link className="link" to="/admin" style={{ width: '100%' }}>
							<LoginButton
								title="Login"
								onClick={() => {}}
								style={{ width: '100%', fontSize: '16px' }}
							/>
							{/* Login
							</LoginButton> */}
						</Link>
					</div>
				</div>
			</div>
		);
	}
}
