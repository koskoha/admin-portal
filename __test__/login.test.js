import * as React from 'react';
import '@babel/polyfill';
import { mount } from 'enzyme';
import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import { BrowserRouter as Router } from 'react-router-dom';

import Login, { LOGIN_USER_MUTATION } from '../src/components/login/Login';

describe('Test case for testing login', () => {
	let wrapper;

	it('renders and matches snapshot', async () => {
		expect(toJSON(wrapper.find('form'))).toMatchSnapshot();
	});

	beforeEach(() => {
		wrapper = mount(
			<MockedProvider>
				<Router>
					<Login />
				</Router>
			</MockedProvider>
		);
	});

	afterEach(() => {
		wrapper.unmount();
	});

	it('username can input email ', () => {
		wrapper.find('input[name="email"]').simulate('change', { target: { value: 'userEmail' } });
		wrapper.update();
		expect(wrapper.find('input[name="email"]').prop('value')).toEqual('userEmail');
		// expect(wrapper.state('email')).toEqual('userEmail');
	});

	it('user can input password ', () => {
		wrapper.find('input[name="password"]').simulate('change', { target: { value: '123456' } });
		wrapper.update();
		expect(wrapper.find('input[name="password"]').prop('value')).toEqual('123456');
	});

	it('login user with correct data', async () => {
		const user = { email: 'kos.koha@gmail.com', password: '123456' };
		const mocks = [
			{
				request: {
					query: LOGIN_USER_MUTATION,
					variables: {
						email: user.email,
						password: user.password
					}
				},
				result: {
					data: {
						login: {
							accessToken: 'any'
						}
					}
				}
			}
		];
		console.log(mocks);
		wrapper.setState({ email: 'kos.koha@gmail.com', password: 123456 });
		Router.router = { replace: jest.fn() };
		wrapper.find('.button').simulate('click');
		await wait(50);
		expect(Router.router.replace).toHaveBeenCalled();
		expect(Router.router.push).toHaveBeenCalledWith({ pathname: '/dashboard/accounts' });
	});
});
