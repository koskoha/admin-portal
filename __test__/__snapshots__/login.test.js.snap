// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test case for testing login renders and matches snapshot 1`] = `null`;
